Migrate to connect-es

edaniels opened the pull request:
Hey all, we talked about doing this migration in #frontend-friends. I figured it would be best to review this code in this repo first, get it merged, and then move it over to the TS SDK. If we somehow need some patches in the meantime, we can just branch off (or worst case revert).

Connect supports pretty much everything we need (node, react native, browser, and more). We can continue the model of having a `global.VIAM` for setting transports and other platform specific needs.

Before:
```shell
✓ 51 modules transformed.
dist/main.js  663.78 kB │ gzip: 87.59 kB
✓ built in 678ms
```

After:
```shell
✓ 133 modules transformed.
dist/main.js  130.79 kB │ gzip: 31.52 kB
✓ built in 356ms
```

edaniels commented on rpc/examples/echo/node-client/src/index.ts:
```diff
@@ -0,0 +1,108 @@
+import { EchoService } from "./gen/proto/rpc/examples/echo/v1/echo_connect.js";
+import { EchoBiDiRequest, EchoMultipleRequest, EchoRequest } from "./gen/proto/rpc/examples/echo/v1/echo_pb.js";
+
+import { createGrpcTransport } from '@connectrpc/connect-node';
```
@njooma this has webrtc "just work" but as far as the TS SDK, I feel like this is kind of ugly and warrants its own node sdk package, unless viam is okay with everyone needing to do this in node. there could be a helper that we may be able to get away with using dynamic imports (that you await)

maximpertsov replied:
no opinion on whether this should be it's own SDK package or not, but it's nice that node is supported in _some_ way with this change

edaniels commented on proto/rpc/examples/fileupload/v1/fileupload.proto:
```diff
@@ -16,7 +16,5 @@ message UploadFileResponse {
-  // Due to an issue described by https://github.com/improbable-eng/ts-protoc-gen/pull/264
-  // we use a streaming response but only expect one response.
-  rpc UploadFile(stream UploadFileRequest) returns (stream UploadFileResponse) {}
+  rpc UploadFile(stream UploadFileRequest) returns (UploadFileResponse) {}
```
Yeah so this was actually wrong and it was never bidi

edaniels commented on buf.lock:
```diff
@@ -4,4 +4,4 @@ deps:
   - remote: buf.build
     owner: googleapis
     repository: googleapis
-    commit: 62f35d8aed1149c291d606d958a7ce32
```
you'll see some buf/goog changes to get us to latest build artifacts/tools

edaniels commented on rpc/examples/echo/frontend/src/index.ts:
```diff
@@ -57,101 +60,90 @@ async function getClients() {
-		pResolve = resolve;
-		pReject = reject;
-	});
+	const clientStream = createWritableIterable<PartialMessage<EchoBiDiRequest>>();
```
Look around here for how to use connect-es. IMO, a lot more ergonomic. Even the bidi case isn't that nasty

edaniels commented on rpc/js/tsconfig.json:
```diff
@@ -7,7 +7,8 @@
     "verbatimModuleSyntax": false,
-    "lib": ["ESNext", "DOM", "DOM.iterable"]
+    "lib": ["ESNext", "DOM", "DOM.iterable"],
+    "exactOptionalPropertyTypes": false,
```
needed for connect rpc passing around optionals

edaniels replied:
I'm gonna try but not hopeful :)

edaniels replied:
figured it out but it was nasty and i love it

maximpertsov commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -38,41 +50,32 @@ export class ClientChannel extends BaseChannel {
-    const id = stream.getId();
-    const activeStream = this.streams[id];
+    const { id } = stream;
+    const activeStream = this.streams[Number(id)];
```
```suggestion
    const activeStream = this.streams[id];
```

Not sure how likely we are to hit a stream count where this would matter, but maybe we should just use `BigInt` throughout? My understanding is that coercing between `BigInt` and `Number` could lead to precision loss per [MDN docs]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#description):

> Be careful coercing values back and forth, however, as the precision of a BigInt value may be lost when it is coerced to a Number value.

edaniels replied:
The proto changed this to a bigint so we could also just change the property on the map?

maximpertsov replied:
> change the property on the map

yep let's do that

edaniels replied:
so bigint cannot be used as an index which makes sense, so how about the record key be a string and we go bigint->string on demand? it has a cost but I don't think we're in that realm of performance needs just yet

maximpertsov replied:
that works too

maximpertsov reviewed (commented):
This is awesome!

I have [one minor comment](https://github.com/viamrobotics/goutils/pull/354/files#r1764946233) related to BigInt/Number coercion. Otherwise LGTM.

mcous reviewed (commented):
Very excited about this work, it's a huge piece of nagging tech debt off our shoulders!

I've done an initial review pass to highlight some areas of confusion for me that I'd like to focus on in our in-person review tomorrow. Also tried to highlight areas where I think extra cleanup, testing, and/or comments would be helpful, either in this PR or in the future, to ensure we're going to be able to maintain this moving forward.

If it can be practically done for this PR, we've been doing a lot of work in Fleet and Viz to use PR/commit stacking to break up large bits of work into smaller chunks. If you're down, I think splitting the updates to `examples` into their own PRs might make this easier to review, understand, and land. I also wonder if there's a way we could really minimize the examples changes to get a better sense of just how much of a "breaking" change this is - as opposed to opportunistic cleanup of the examples. (I think example cleanup is valuable but could happen in follow-up PRs by other contributors to gain familiarity.)

ethanlook commented on rpc/js/src/dial.ts:
```diff
@@ -339,15 +353,15 @@ export async function dialWebRTC(
   // TODO(RSDK-2836): In general, this logic should be in parity with the golang implementation.
   // https://github.com/viamrobotics/goutils/blob/main/rpc/wrtc_client.go#L160-L175
   const config = await getOptionalWebRTCConfig(signalingAddress, host, opts);
```
Looks like we're doing this to get additional ice servers from the machine. Why is this necessary?

edaniels replied:
added a comment!

ethanlook commented on rpc/js/src/errors.ts:
This error is used throughout the TS SDK - what's the rationale for removing it?

edaniels replied:
It wasn't used in this package but I'll move it back

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -90,180 +96,185 @@ export interface Credentials {
-      TransFact = grpc.CrossBrowserHttpTransport;
-    }
-    return TransFact({ withCredentials: false })(opts);
+  let transFact: TransportFactory;
+  try {
+    transFact = window.VIAM.GRPC_TRANSPORT_FACTORY;
+  } catch {
+    transFact = createGrpcWebTransport;
+  }
```
yep this is better

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -412,64 +400,58 @@ export async function dialWebRTC(
+            'rpc-host': host,
+          },
+        });
+      } catch (err) {
+        console.error(err);
+      }
```
left a comment on it explaining why it's okay

edaniels commented on rpc/examples/echo/frontend/src/index.ts:
```diff
@@ -17,15 +20,15 @@ declare global {
-  inner.setAttribute("data-testid", "message");
-  inner.innerText = text;
-  elem.appendChild(inner);
+	const selector = `[data-testid="${type}-${method}"]`;
```
this file is actually inconsistent on spacing/tabs so I think I'm going to leave this unless you really care

mcous replied:
Lol of course. No, happy to let prettier deal with this later

edaniels commented on rpc/examples/echo/frontend/src/index.ts:
```diff
@@ -57,101 +60,90 @@ async function getClients() {
+		const webRTCConn = await dialWebRTC(thisHost, webrtcHost, opts);
+		const webrtcClient = createPromiseClient(EchoService, webRTCConn.transport);
+		await renderResponses(webrtcClient, "wrtc");
+	} catch (err) {
+		console.error("error trying WebRTC", err);
+	}
```
yep

edaniels commented on rpc/examples/echo/node-client/src/index.ts:
```diff
@@ -0,0 +1,108 @@
+import { createGrpcTransport } from '@connectrpc/connect-node';
+import wrtc from "node-datachannel/polyfill";
+global.window = global.window || {};
+window.VIAM = {
+	GRPC_TRANSPORT_FACTORY: (opts: any) => createGrpcTransport({ httpVersion: "2", ...opts }),
+};
```
it's more for the isomorphism. there's likely a better pattern but I don't want to spend the cycles on that right now since it's for lesser used use cases

edaniels replied:
will do globalThis

edaniels commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -38,41 +50,32 @@ export class ClientChannel extends BaseChannel {
-        new Uint8Array(event.data as ArrayBuffer)
-      );
+      resp = Response.fromBinary(new Uint8Array(event.data as ArrayBuffer));
     } catch (e) {
       console.error('error deserializing message', e);
       return;
     }
```
sketchy indeed. will remove

edaniels commented on rpc/js/src/ClientStream.ts:
```diff
@@ -18,77 +38,285 @@ import {
+        contextValues: contextValues ?? createContextValues(),
+        message,
+      },
+      next: async (req: UnaryRequest<I, O>): Promise<UnaryResponse<I, O>> => {
```
refactor incoming

edaniels commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -82,81 +85,145 @@ export class ClientChannel extends BaseChannel {
 
   private nextStreamID(): Stream {
     const stream = new Stream();
```
did this everywhere as much as I could

edaniels commented on rpc/js/src/ClientStream.ts:
```diff
@@ -98,158 +326,134 @@ export class ClientStream extends BaseStream implements grpc.Transport {
+  const result = new Metadata({
+    md: {},
+  });
+  const { md } = result;
+  headers.forEach((key, value) => {
     const strings = new Strings();
-    strings.setValuesList(values);
-    md.set(key, strings);
+    strings.values = [value];
+    md[key] = strings;
   });
-  if (result.getMdMap().getLength() === 0) {
+  if (Object.keys(result.md).length === 0) {
     return undefined;
   }
   return result;
 };
```
not adding tests here for both but put in your nice changes

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -90,180 +96,211 @@ export interface Credentials {
+    message: PartialMessage<I>,
+    contextValues?: ContextValues
+  ): Promise<UnaryResponse<I, O>> {
+    let newHeaders = cloneHeaders(header);
```
I'll keep a lookout but also once we start linting

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -664,13 +591,78 @@ export async function dialWebRTC(
+          if (response.stage.case == 'init') {
+            if (haveInit) {
+              await sendError('got init stage more than once');
+              continue;
```
fixed

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -664,13 +591,78 @@ export async function dialWebRTC(
-    await clientEnd;
-    await cc.ready;
+    const processCallResponses = async () => {
+      try {
```
will address in refactor

edaniels commented:
did a bunch of cleaning. next step is a signaling refactor and a comments pass if needed. then ready for re-review. @mcous @ethanlook feel free to take a look at the ClientStream/UnaryClientStream/StreamClientStream refactor. I think it's a lot easier to read now and it minimizes the complexity of the new Promise logic as minimally as I could.

edaniels commented:
@ethanlook @mcous ready for another review. since the last comment, I refactored signaling as requested. It's still complicated, but a little easier to read. Also split the commits up.

edaniels commented on rpc/dialer.go:
```diff
@@ -107,6 +107,7 @@ func (cd *cachedDialer) DialDirect(
 	opts ...grpc.DialOption,
 ) (ClientConn, bool, error) {
 	return cd.DialFunc("grpc", target, keyExtra, func() (ClientConn, func() error, error) {
+		//nolint:staticcheck
```
@dgottlieb heads up for rdk and other go users of grpc that grpc is removing DialContext in favor of NewClient + Connect. I'm just ignoring these warnings here for now, but these may pop up in rdk.

edaniels commented:
Also made some changes to the buf yaml to reduce how many files this PR changes. There was an unneeded version upgrade.

mcous commented on rpc/js/src/BaseChannel.ts:
```diff
@@ -1,4 +1,4 @@
-import type { ProtobufMessage } from '@improbable-eng/grpc-web/dist/typings/message';
+import { Message } from '@bufbuild/protobuf';
```
Import only used as type

```suggestion
import type { Message } from '@bufbuild/protobuf';
```

edaniels replied:
shouldn't all these kinds of things be picked up by a linter?

mcous replied:
Yup, see https://github.com/viamrobotics/goutils/pull/354#discussion_r1771674359

mcous commented on rpc/js/src/ClientStream.ts:
```diff
@@ -1,7 +1,13 @@
-import { grpc } from '@improbable-eng/grpc-web';
+import {
```
```suggestion
import type {
```

mcous commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -1,6 +1,18 @@
-import type { grpc } from '@improbable-eng/grpc-web';
+import {
```
```suggestion
import type {
```

mcous commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -1,6 +1,18 @@
+  PartialMessage,
+  ServiceType,
+} from '@bufbuild/protobuf';
+import {
```
```suggestion
import type {
```

mcous commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -1,6 +1,18 @@
+} from '@connectrpc/connect';
 import { BaseChannel } from './BaseChannel';
-import { ClientStream } from './ClientStream';
+import { ClientStream, ClientStreamConstructor } from './ClientStream';
```
```suggestion
import type { ClientStream, ClientStreamConstructor } from './ClientStream';
```

mcous commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -38,41 +52,26 @@ export class ClientChannel extends BaseChannel {
-      console.error('error deserializing message', e);
-      return;
-    }
+    let resp = Response.fromBinary(new Uint8Array(event.data as ArrayBuffer));
```
Do we need a runtime check or assert for this type cast to `as ArrayBuffer`? Or should `MessageEvent<any>` be defined more strictly?

edaniels replied:
typed it more

mcous commented on rpc/js/src/ClientChannel.ts:
```diff
@@ -38,41 +52,26 @@ export class ClientChannel extends BaseChannel {
     }
   }
 
   private onChannelMessage(event: MessageEvent<any>) {
```
Should this be `MessageEvent<I>` (or `MessageEvent<O>` if I'm misunderstanding the terminology here) rather than `any`?

edaniels replied:
it's neither. this is a message coming from a data channel which will be bytes

mcous commented on rpc/js/src/main.ts:
```diff
@@ -1,7 +1,10 @@
+import { Transport } from '@connectrpc/connect';
```
```suggestion
import type { Transport } from '@connectrpc/connect';
```

mcous commented on rpc/js/src/main.ts:
```diff
@@ -1,7 +1,10 @@
-    VIAM: any;
-  }
+  // eslint-disable-next-line vars-on-top,no-var
+  var VIAM: {
+    GRPC_TRANSPORT_FACTORY: (opts: any) => Transport;
+  };
```
Ensure these things are typed as potentially not existing

```suggestion
  var VIAM:
    | {
        GRPC_TRANSPORT_FACTORY?: (opts: any) => Transport;
      }
    | undefined;
```

mcous commented on rpc/examples/echo/node-client/src/index.ts:
```diff
@@ -0,0 +1,104 @@
+
+import { createGrpcTransport } from '@connectrpc/connect-node';
+import wrtc from "node-datachannel/polyfill";
+declare global {
+	var VIAM: {
+		GRPC_TRANSPORT_FACTORY: (opts: any) => Transport;
+	};
+}
```
Not needed, since it's declared in `main.ts` of `@viamrobotics/rpc`
```suggestion

```

mcous commented on rpc/js/tsconfig.json:
```diff
@@ -1,7 +1,7 @@
   "compilerOptions": {
-    "target": "ES5",
+    "target": "esnext",
     "useDefineForClassFields": true,
     "importHelpers": true,
```
Some suggested TSConfig changes:

- Remove `useDefineForClassFields` and `importHelpers` as unnecessary
- Add `"noEmit": true`
    - Can then remove `--no-emit` from `scripts.typecheck` in `package.json`
- Remove `"verbatimModuleSyntax": false` to use our base default of `true`
    - This will catch all the `import type { ... };` problems, a few of which I pointed out
- Add `"noEmit": false` to `tsconfig.build.json` to keep the `.d.ts` file generation working

edaniels replied:
didn't seem to catch `import { Message } from '@bufbuild/protobuf';` btw

mcous replied:
Yeah, I've seen it miss some stuff in the app, too. As long as we get the bulk of them that's good.

I think `vite` throws some build warnings about unused imports, too (haven't looking into how to promote those to errors, yet) and I suspect newer versions of `@typescript-eslint` might do a better job of detecting this, too, so hopefully we can get full coverage once this is over in the TS SDK repo

mcous reviewed (commented):
Still working through this, but here's a quick batch of housekeeping comments just to keep things moving

mcous commented on rpc/js/src/signaling-exchange.ts:
```diff
@@ -0,0 +1,391 @@
+      this.totalCallUpdateDuration += callUpdateDuration;
+    } catch (error) {
+      if (
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        this.exchangeDone ||
```
[Not a blocker, not the best time to resolve] Same comment here: it looks a lot like `this.exchangeDone` is already checked above, and I don't immediately see any method calls that would change `exchangeDone` between the two checks. Is something coming in externally and causing it to change?

Either way, I think if it's enough to confuse ESLint, it's enough to confuse a future reader like me.

edaniels replied:
An await is used prior to exchange done bring checked which means based on how the event loop orders things, it may be set to true from another function. This is a case of unseen concurrency in JavaScript

edaniels replied:
Without this extra check, you'll see a needs atomic check lint and the two resolutions to the lints are in conflict. I've seen this several times for concurrent code in JS and it's frustrating to work around.

edaniels replied:
managed to simplify the code since there's no more factories here really

mcous reviewed (approved):
I think this is a great state for a testing release! Appreciate all your work on this

edaniels commented on rpc/js/src/dial.ts:
```diff
@@ -90,641 +97,425 @@ export interface Credentials {
       }
 
-      accessToken = thisAccessToken;
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- await race
```
Same response as the other due to await

edaniels replied:
managed to simplify the code since there's no more factories here really

ethanlook commented on rpc/js/src/SignalingExchange.ts:
Nice, this encapsulation is really helpful for understanding the signaling exchange and how it fits in the bigger picture of dialing. Much appreciated!

ethanlook reviewed (approved):
Looks great, thanks for the lint cleanup as well! And thank you again for walking us through this code.
